cmake/macros/CompositeSources.cmake

```diff
@@ -47,35 +44,69 @@ function(composite_sources target sources_var)
     return()
   endif()
 
+  # Get rid of the periods in the extensions and create each list of composite
+  # sources.
+  set(extensions "")
+  foreach(extension ${COMPOSITE_SOURCE_EXTENSIONS})
+    string(MAKE_C_IDENTIFIER extension ${extension})
+    list(APPEND extensions ${extension})
+    set(sources_${extension} "")
+  endforeach()
+
+  # Sort each source file into a list.
+  foreach(source ${sources})
+    get_filename_component(extension "${source}" EXT)
+    string(MAKE_C_IDENTIFIER extension ${extension})
+    list(APPEND sources_${extension} "${source}")
+  endforeach()
+
+  set(sorted_sources "")
+  foreach(extension ${extensions})
+    list(APPEND sorted_sources ${sources_${extension}})
+  endforeach()
+
   set(composite_files "")
   set(composite_sources "")
 
+  list(GET sorted_sources 0 first_source)
+  get_filename_component(first_source_ext "${first_source}" EXT)
+  set(composite_ext ${first_source_ext})
+
   while(num_sources GREATER 0)
     # Pop the first element
-    list(GET sources 0 source)
-    list(REMOVE_AT sources 0)
-    list(LENGTH sources num_sources)
+    list(GET sorted_sources 0 source)
+    list(REMOVE_AT sorted_sources 0)
+    list(LENGTH sorted_sources num_sources)
 
     # Check if we can safely add this to a composite file.
     get_source_file_property(generated "${source}" GENERATED)
     get_source_file_property(is_header "${source}" HEADER_FILE_ONLY)
     get_source_file_property(skip_compositing "${source}" SKIP_COMPOSITING)
     get_filename_component(extension "${source}" EXT)
 
+    # Get the next source file's extention
+    if(num_sources GREATER 0)
+      list(GET sorted_sources 0 next_source)
+      get_filename_component(next_extension "${next_source}" EXT)
+    else()
+      set(next_extension "")
+    endif()
+    
     if(NOT generated AND NOT is_header AND NOT skip_compositing AND
```

Comment:
Adding this logic to the pre-sorting step may actually make the main loop easier to understand. Perhaps this refactor is a good opportunity to do that? :D